best_practices.md

```diff
@@ -261,3 +261,96 @@ Including `using namespace` in header files can lead to namespace pollution and
 - **Code Clarity**: Improves code clarity by explicitly specifying the namespace.
 - **Maintainability**: Enhances maintainability by avoiding unintended interactions between different parts of the code.
 
+## 14. Avoid Bool Arguments in APIs
+
+### Practice
+Avoid using plain bool arguments in APIs, especially when the meaning isn't clear at the call site. Instead, use enum class to provide context.
+
+### Explanation
+Using a bool argument in a function call can make the code less readable and harder to maintain, as the purpose of the true or false value is often unclear. By using an enum class, you make the intention explicit, improving code clarity.
+
+### Example
+Avoid:
+```cpp
+tensor = tt::tt_metal::tilize_with_val_padding(tensor, output_shape, 0, output_memory_config, dtype, true);
+```
+Prefer:
+```cpp
+enum class ThreadingOption { SingleCore, MultiCore };
+tensor = tt::tt_metal::tilize_with_val_padding(tensor, output_shape, 0, output_memory_config, dtype, ThreadingOption::MultiCore);
+```
+Also consider giving enums power-of-2 values to pass them all as a single argument, e.g. 
+```cpp
+Options::FOO | Options::BAR
+```
+
+### Motivation
+- **Readability:** Enhances readability by making the purpose of the argument explicit.
+- **Maintainability:** Reduces the likelihood of misusing the API, making it easier to maintain and extend.
+- **User Experience:** Helps in winning over users by providing a clearer, more intuitive API.
+
+## 15. Initialize Primitive Types on Declaration
+### Practice
+Always initialize primitive types (e.g., size_t, int, float, bool, pointers) at the point of declaration.
+
+### Explanation
+Forgetting to initialize primitive types can lead to unpredictable behavior, as uninitialized variables may contain garbage values. This can cause hard-to-debug issues, especially in large codebases.
+
+### Example
+Avoid:
+```cpp
+struct PadDimension {
+    std::size_t front;
+    std::size_t back;
+
+    static constexpr auto attribute_names = std::make_tuple("front", "back");
+    const auto attribute_values() const { return std::make_tuple(std::cref(this->front), std::cref(this->back)); }
+};
+```
+Prefer:
```

Comment:
One caveat here is that you lose the `std::is_trivially_default_constructible` property of the type. So in an unlikely case when that becomes important, you may prefer the original version.